### ISSUE TITLE
Reload image after a failed load in Jetpack Compose

### ISSUE THREAD

**Issue by FishHawk** [coil-kt/coil#884]
**Is your feature request related to a problem? Please describe.**
```kotlin
val painter = rememberImagePainter(url)
Image(
    modifier = Modifier.fillMaxSize(),
    painter = painter,
    contentDescription = null,
    contentScale = ContentScale.Fit
)
when (val state = painter.state) {
    is ImagePainter.State.Error -> {
        TextButton(onClick = { }) { Text("retry") }
    }
}
```
After a failed load, the user should be able to reload the image via the retry button.

**Describe the solution you'd like**
It is better to provide `reload` method, but it seems that setting `request` in `ImagePainter` to public would also work.

**Comment by colinrtwhite** [comment]
Still figuring out a good public API for this, but if you need this today you can force retry by changing a parameter:

```kotlin
var retryHash = 0
val painter = rememberImagePainter(
    request = ImageRequest.Builder(LocalContext.current)
        .data(url)
        .setParameter("retry_hash", retryHash, cacheKey = null)
        .build()
)
Image(
    modifier = Modifier.fillMaxSize(),
    painter = painter,
    contentDescription = null,
    contentScale = ContentScale.Fit
)
when (val state = painter.state) {
    is ImagePainter.State.Error -> {
        TextButton(onClick = { retryHash++ }) { Text("retry") }
    }
}
```

Currently, I'm thinking the public API should be something like this, but let me know what you think!

```kotlin
var requestHandle = ImageRequestHandle()
val painter = rememberImagePainter(
    request = ImageRequest.Builder(LocalContext.current)
        .data(url)
        .requestHandle(requestHandle)
        .build()
)
Image(
    modifier = Modifier.fillMaxSize(),
    painter = painter,
    contentDescription = null,
    contentScale = ContentScale.Fit
)
when (val state = painter.state) {
    is ImagePainter.State.Error -> {
        TextButton(onClick = { requestHandle.restart() }) { Text("retry") }
    }
}
```

**Comment by FishHawk** [comment]
Thanks! I'll try it later.

The api is indeed a problem. Your example is very similar to `focusRequester`. I'm ok with it. But since there must be an ImagePainter object here, I think it would be simpler to have an ImagePainter with a retry method.

**Comment by colinrtwhite** [comment]
Yep, adding a method to `ImagePainter` might be ok (and more discoverable), though I think it might not work well with the new [`AsyncImage`](https://github.com/coil-kt/coil/blob/main/CHANGELOG.md#200-alpha04---november-22-2021) component that'll be added in Coil 2.0 since the `ImagePainter` is only exposed in `AsyncImageScope`. Having the retry handler as part of the request works for both `ImagePainter` and `AsyncImage`.

**Comment by FishHawk** [comment]
I see. In that case, I would prefer code like this:
``` kotlin
AsyncImage(...) { state ->
    if (state is AsyncImagePainter.State.Error) {
        TextButton(onClick = { painter.retry() }) { Text("retry") }
    } else {
        AsyncImageContent()
    }
}
```

If someone needs to use `retry` outside of `AsyncImageScope`, he can wrap a lambda himself: 
`val retryHandle = { painter.retry() }`
(I don't think this is a good pattern, but some global components like `BottomSheet` do cause this situation. I think this is Jetpack Compose's design mistake.)

If someone needs to refresh multiple images at once, he can use a `Flow<SideEffect>` just like the mvi architecture and collect it inside `AsyncImageScope`.

Anyway, I prefer the direct api, mainly because `focusRequester` messed up my code.

**Comment by sarseneaultrp** [comment]
A retry handle would be desirable. @colinrtwhite - Is this feature something you see in a near future release?

**Comment by colinrtwhite** [comment]
No immediate plans for adding this (currently focused on making sure `AsyncImage` is solid), though it'll likely be in the final 2.0 release or 2.1.

**Comment by BenjyTec** [comment]
Are there any updates on this issue, or a workaround?

**Comment by trOnk12** [comment]
would be nice to add a support for this

**Comment by NasiaKoutsopoulou** [comment]
is this implemented and exposed to the API?
if not, is there any estimation on when it will be available?

**Comment by eecs441staff** [comment]
Only works for me if `.setParameter("retry_hash", retryHash, memoryCacheKey = null)` in @colinrtwhite 's solution is used without the `memoryCacheKey` argument, thus `.setParameter("retry_hash", retryHash)`.

Also if `retryHash` is a Boolean, the model toggles between its first two instances, so it does need to have a range.

**Comment by PythonVader** [comment]
I am getting this Error when i add an image URI with an "i". 

Heres the Code:
AsyncImage(
                    rememberAsyncImagePainter(my_image_uri))

Getting this Runtime Error if this helps:

ava.lang.NoClassDefFoundError: Failed resolution of: Landroidx/compose/runtime/PrimitiveSnapshotStateKt;
                                                                                                    	at coil.compose.AsyncImagePainter.<init>(AsyncImagePainter.kt:166)
                                                                                                    	at coil.compose.AsyncImagePainterKt.rememberAsyncImagePainter-5jETZwI(AsyncImagePainter.kt:141)

**Comment by massivemadness** [comment]
Any updates? 👀

**Comment by PythonVader** [comment]
> Any updates? 👀

Yes, The issue was due to an incompatibility issue(with the gradle or compose version I can't remember). I have downgraded the coil version suddenly it works.

So to solve this I have checked the coils version compatibility with my project dependencies.   Now I'm using the latest version and works fine.

**Comment by colinrtwhite** [comment]
Hi folks, quick update on this. This is definitely something I want to address properly in Coil 3.0. Ideally, we can kill two birds with one stone and hoist `AsyncImagePainter` so that way users can also observe its other properties (like `state` or `request`). For example:

```kotlin
val painter = rememberAsyncImagePainter(
    model = "https://example.com/image.jpg"
)

AsyncImage(
    model = painter,
    contentDescription = null,
)

when (state) {
    is AsyncImagePainter.State.Error -> {
        ErrorButton(onClick = { painter.restart() })
    }
}
```

There are some implementation details that might cause this to not work in practice, but it's the API I'd like!

**Comment by brinsche** [comment]
Related to this I'd love to have the ability to disable "autostarting" the request, I recently had very slow connections speeds and it would have been nice to not automatically trigger some image loads.

I assume I could wrap those images and instantly fail them with an interceptor but I think it would be even nicer UX to be able to differentiate the initial request from retries. Thanks for all the work so far!

**Comment by brinsche** [comment]
Sure, but I'd want to show the same placeholder and reload due to error and reload due to non autostart are conceptually similar in my mind, but it's obviously fine if coil doesn't want to take on the api burden for such a niche use case.

**Comment by Andrew0000** [comment]
Maybe not the optimal solution from the recomposition perspective but seems like we can use the .key() function to easily "reload" the AsyncImage in Coil 3.

Something like:

```
val failure = remember(url) { mutableIntStateOf(0) }
val attempt = remember(url) { mutableIntStateOf(0) }
LaunchedEffect(failure.intValue) {
    val delay = (failure.intValue * 1_000L).coerceAtMost(10_000L)
    delay(delay)
    attempt.intValue = failure.intValue
}

key(attempt.intValue) {
    AsyncImage(
        ...
        onState = { state: AsyncImagePainter.State ->
            if (state is AsyncImagePainter.State.Error) {
                failure.intValue++
            }
        },
    )
}
```

**Comment by colinrtwhite** [comment]
This supported via `AsyncImagePainter.restart()` in Coil `3.0.0-rc01`. It avoids recomposing too so it should be more efficient than using [the previous solution](https://github.com/coil-kt/coil/issues/884#issuecomment-975932886).

**Comment by Jeanno** [comment]
Very cool. Thanks

**Comment by colinrtwhite** [comment]
Closing this out as `AsyncImagePainter.restart` is available in `3.0.0`. If you want the benefits of `AsyncImage` (downsampling images to the size of the constraints), but still want to use this API, check out the section using `rememberConstraintsSizeResolver` [here](https://coil-kt.github.io/coil/compose/#rememberasyncimagepainter).
